lib/utils.js

```diff
@@ -7,98 +10,79 @@ var assert = require("assert");
-  assert.ok(databaseName !== undefined);
-
-  //Parsing options
-  if (options) {
-    var parsedOptions = options.split(",");
-    //If options are parsed, they should have an even number of elements in the split array
-    assert.ok(parsedOptions.length > 0);
-
-    for (var i = 0; i < parsedOptions.length; i++) {
-      var splitOption = parsedOptions[i].split("=");
-      //Each option should be something=value
-      assert.ok(splitOption && splitOption.length === 2);
-      assert.ok(splitOption[0].length > 0);
-      assert.ok(splitOption[1].length > 0);
-      //Checking it does not already exist
-      assert.ok(!result.database.driver_options[splitOption[0]]);
-
-      result.database.driver_options[splitOption[0]] = splitOption[1];
-    }
+    return pathname ? pathname.replace('/', '') : pathname;
   }
 
-  var host_port;
-  if (hostList.length > 1) {
-    result.database.host = [];
-    result.database.port = [];
-    for (i = 0; i < hostList.length; i++) {
-      host_port = hostList[i].split(":");
-      assert.ok(host_port[0] && host_port[0].length > 0);
-      result.database.host.push(host_port[0]);
-      if (host_port.length > 1) {
-        result.database.port.push(host_port[1]);
-      } else {
-        result.database.port.push('27017');
-      }
+  // Parse the query string from the last host portion
+  function getOptionsObject () {
+    if (parts.length > 1) {
+      // Options should be appended to the last hosts querystring
+      return qs.parse(url.parse(parts[parts.length - 1]).query);
     }
+
+    return {};
   }
-  else {
-    host_port = hostList[0].split(":");
-    assert.ok(host_port[0] && host_port[0].length > 0);
-    result.database.host = host_port[0];
-    if (host_port.length > 1) {
-      result.database.port = host_port[1];
+  
+  // Authentication is defined in the primary host URL
+  function getAuthenticationObject () {
+    const authString = url.parse(parts[0]).auth;
+
+    if (authString) {
+      return {
+        source: getName(),
+        user: authString.split(':')[0],
+        pass: authString.split(':')[1]
+      };
     } else {
-      result.database.port = '27017';
+      return {};
     }
   }
 
+  // Only the primary host string contains the "mongodb://" so we use this
+  // function to prepend it to others when passing to url.parse
+  function prependProtocol (str) {
+    if (str.indexOf('mongodb://') === -1) {
+      return 'mongodb://' + str;
+    } else {
+      return str;
+    }
+  }
 
-  //Verify the config to be returned
-  assert.ok(result.database.name.length > 0);
-  assert.ok(result.database.auth.user.length > 0);
-  assert.ok(result.database.auth.pass.length > 0);
-
-
-  return result;
+  var result = {
+    database: {
+      driver_options: getOptionsObject(),
+      auth: getAuthenticationObject(),
+      name: getName(),
+      
+      // Use map since we can have multiple hosts and ports
+      host: parts.map((p) => url.parse(prependProtocol(p)).hostname),
```

Comment:
@rachael-oregan - we only support node.js v4+ now, should this not be ok? Will remove, but possibly worth highlighting?